url2b4a2d655b_Yehonal_unisa_prog2_jdm_tgz-pJ8-com_jdeals_tests_TestUserBuyingJ8/src/com/jdeals/main/controller/Store.java

```diff
@@ -0,0 +1,446 @@
+                    lastId = i.getId();
+                }
+            }
+
+            Item.nextId.set(lastId);
+
+            this.items = tmp;
+        }
+
+        this.discounts = storer.loadList("database/discounts.db");
+    }
+
+    /**
+     * Add item to store.
+     *
+     * @param item instance of the Item
+     * @throws Exception can throws an exception if id already exists
+     * @see Item
+     */
+    public void addItem(Item item) throws Exception {
+        if (this.getItem(item.getId()) != null) {
+            throw new Exception("An item with same id already exists!");
+        }
+
+        items.add(item);
+    }
+
+    /**
+     * Deletes item from store.
+     *
+     * @param id item id
+     * @return true, if successful
+     * @see Item
+     */
+    public boolean delItem(int id) {
+        Item i = this.getItem(id);
+        return this.items.remove(i);
+    }
+
+    /**
+     * Gets item from store.
+     *
+     * @param id item id
+     * @return the item
+     * @see Item
+     */
+    public Item getItem(int id) {
+        for (Item i : this.items) {
+            if (i.getId() == id) {
+                return i;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Gets instance of JDealsController.
+     *
+     * @return JDealsController
+     */
+    public JDealsController getMainCtrl() {
+        return mainCtrl;
+    }
+
+    /**
+     * Adds discount to store.
+     *
+     * @param d DiscountInterface
+     * @throws Exception can throw an exception in case a rule with same param.
+     * already exists
+     * @see DiscountInterface
+     */
+    public void addDiscount(DiscountInterface<Object> d) throws Exception {
+        if (this.discounts.contains(d)) {
+            throw new Exception("This rule already exists");
+        }
+
+        this.discounts.add(d);
+    }
+
+    /**
+     * Removes discount from the store.
+     *
+     * @param indexId the index id
+     * @see DiscountInterface
+     */
+    public void removeDiscount(int indexId) {
+        this.discounts.remove(indexId);
+    }
+
+    /**
+     * Gets discount list.
+     *
+     * @return ArrayList<DiscountInterface> stored discounts
+     * @see DiscountInterface
+     */
+    public ArrayList<DiscountInterface<Object>> getDiscounts() {
+        return this.discounts;
+    }
+
+    /**
+     * Add item to cart.
+     *
+     * @param item the item
+     * @param quantity the quantity
+     * @return true, if successful
+     * @throws Exception the exception
+     * @see Item
+     * @see Order
+     * @see User
+     */
+    public boolean addItemToCart(Item item, int quantity) throws Exception {
+        User cUser = this.mainCtrl.getCurUser();
+        Order cOrder = cUser.getCurOrder();
+
+        checkQuantity(item, cOrder, quantity, true);
+
+        cOrder.addItem(item, quantity);
+
+        try {
+            this.discountOrderCost(cOrder);
+        } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)